### DISCUSSION TITLE
Skill Score Diagram for Rapid Intensification

### DISCUSSION THREAD

**Discussion by YanJin-NOAA** [dtcenter/METplus#1244] [Statistical Computation] [answered]
Hi!

I wonder if MET-TC tool can plot the skill score of the rapid intensification like this figure?

![image](https://user-images.githubusercontent.com/79869112/140784358-0f68fbe8-c067-4ba7-ae69-83fc8efafb20.png)

I found that MET-Stat tool can compare the occurrence of rapid intensification between adeck and bdeck, and output a table. So I wonder if I can use MET to plot the figure? Or I need to write a script to plot the figure by myself?

Thanks very much.
Best,
Yan

**Comment by JohnHalleyGotway** [comment, marked as answer]
@YanJin-NOAA I see you have a question about plotting categorical statistics of Tropical Cyclone rapid intensification events. Thanks for including an image to illustrate your question. This type of plot is called a "Performance Diagram" and it's used to summarize multiple statistics from a 2x2 contingency table on the same plot.

Unfortunately, I believe the answer to your question is no. At this time, there is not an *easy* way to create the plot you're after. But let me tell you about 2 related details:

(1) The METviewer software IS able to create performance diagrams. Looking in the [METviewer GUI](http://dtcenter.ucar.edu/met/metviewer/metviewer1.jsp), the tab named "Perf" is used to create performance diagrams. So if the RIRW contingency table counts lived in the database for METviewer, you could create that plot.

(2) The TC-Stat "[-job RIRW](https://met.readthedocs.io/en/latest/Users_Guide/tc-stat.html#statistical-aspects)" is presumably the job you're running. That's the job that inspects the TC-Pairs output looking for the occurrence of rapid intensification and defining contingency table counts and statistics. However, the contingency table output it writes is a custom format and looks something like this:
```
COL_NAME: ATIME  BTIME  AEXACT BEXACT ATHRESH BTHRESH WINDOW_BEG WINDOW_END TOTAL FY_OY FY_ON FN_OY FN_ON
RIRW_CTC: 240000 240000 true   true   <=-15   <=-15   -120000    120000        42     2     0     2    38
```
While all the information is present, it does not follow the full .stat file format. So it can't be easily loaded into METviewer. We didn't initially format it like that because it wasn't obvious how columns like ATIME, BTIME, AEXACT, BEXACT, ATHRESH, BTHRESH, WINDOW_BEG, and WINDOW_END should be mapped into the existing .stat header columns.

Perhaps we should revisit this decision and update TC-Stat to actually write CTC and CTS line types that can be loaded into a database for METviewer?

Hope that helps clarify.

**Reply by YanJin-NOAA** [reply]
Hi John,

Thanks very much. I am thinking that I can use TC-Pairs and TC-Stats to get the data, then use Python to plot the figure. 

Would you mind taking a look at my script to check if I used TC-Paris and TC-Stats correctly? I planned to include HWRF, HMON, HF2A, HF2B, HF2D models in the performance diagram. The TC cases are 2021 TCs over AL, and include the case 02 to case 21. My work directory is on Jet: /mnt/lfs4/HFIP/hwrfv3/Yan.Jin/vrfy_METplus4.0/RI_2021_AL

TC-Pairs config file: TCPairsConfig_run1
The script to run the TC-Pairs: tc_pairs_run1.sh
TC-Stat config file: TCStatConfig_run1
The script to run the TC-Stat: tc_statUseConfig_run1.sh.

However, the -job rirw didn't generate a correct output file. I wonder if the parameters in the TC-Stat are not set correctly. For instance, In 
```
rirw = {
   track  = NONE;       Specify which track types to search (NONE, ADECK,
                        BDECK, or BOTH)
```
I am confused about the parameter for "track" here. 

Many thanks.
Best,
Yan

**Reply by JohnHalleyGotway** [reply]
@YanJin-NOAA, I tried to log onto jet this morning but, of course, I picked the day that it's down for maintenance! Wanted to let you know that I saw this message and will make a note to remind myself to take a look tomorrow.

**Reply by JohnHalleyGotway** [reply]
@YanJin-NOAA, I took a look at this on jet today.
I see that your TC-Pairs output file contains plenty of data:
```
> cat output/tc_pairs_run1_AL2021.tcst | wc -l
150361
```
Next, I ran the following commands to get RIRW contingency table output:
```
cd /mnt/lfs4/HFIP/hwrfv3/Yan.Jin/vrfy_METplus4.0/RI_2021_AL
module load intel
module load intelpython/3.6.5
module load netcdf/4.6.1
module load hdf5/1.10.4
module load contrib
module use /contrib/met/modulefiles
module load met/10.0.0
tc_stat -lookin output/tc_pairs_run1_AL2021.tcst -job rirw -by AMODEL -out ~/rirw_ctc_by_model.txt -v 4
```
Some details to note:
- You can either run a single job directly on the command line (as I've done above) or run it through a config file (as your script does).
- You do not need to use the "-dump_row" option unless you have a good reason for doing so. For each input line read by TC-Stat, if it survives the filtering criteria, then it's written to the output dump_row file. Since you really aren't doing any additional filtering in TC-Stat, I expect that the dump_row will be exactly the same as the input file.
- Using "-by AMODEL" is important. That computes results separately for each input model name (i.e. AMODEL column in the TC-Pairs output). Without that, all results are just lumped together.
- The "-out ~/rirw_ctc_by_model.txt" writes the contingency table output to a file rather than the screen, which is the default. I'm guessing you were getting output, but it was just printed to standard out.
- Running at verbosity level 4 (-v 4) prints log messages like this:
```
DEBUG 4: Found ADECK RI/RW: TrackPairInfo: STORMID = AL062021, ADECK = HMON, BDECK = BEST, INIT = 20210813_060000, NPoints = 71, VALID = 20210816_090000, LEAD = 750000, 240000 exact ADECK change 42 to 77, 35>=30.0
DEBUG 4: Found BDECK RI/RW: TrackPairInfo: STORMID = AL072021, ADECK = HWRF, BDECK = BEST, INIT = 20210811_180000, NPoints = 49, VALID = 20210820_180000, LEAD = NA, 240000 exact BDECK change 50 to 80, 30>=30.0
```
So that's pretty detailed info about each RI event occurring in the forecast or BEST tracks.

The job takes about 3 minutes to run. By default, it writes CTC and CTS output information. Here's the CTC output I see in the ~/rirw_ctc_by_model.txt output file I wrote.
```
COL_NAME: ATIME  BTIME  AEXACT BEXACT ATHRESH BTHRESH WINDOW_BEG WINDOW_END AMODEL TOTAL FY_OY FY_ON FN_OY FN_ON
RIRW_CTC: 240000 240000 true   true   >=30.0  >=30.0  000000     000000     HF2A    4803    32    45   120  4606
RIRW_CTC: 240000 240000 true   true   >=30.0  >=30.0  000000     000000     HF2B    4779    44   101   126  4508
RIRW_CTC: 240000 240000 true   true   >=30.0  >=30.0  000000     000000     HF2D    4116    28    35    91  3962
RIRW_CTC: 240000 240000 true   true   >=30.0  >=30.0  000000     000000     HMON    6330    57    62   170  6041
RIRW_CTC: 240000 240000 true   true   >=30.0  >=30.0  000000     000000     HWRF    6892    84   163   160  6485
```
Note that you can turn on very detailed matched pair output by adding:
```
-out_line_type CTC,CTS,MPR
```
Also note that you can adjust the definition of RI, as well as the time window matching criteria, as described in "Job: RIRW" section of the [MET User's Guide](https://met.readthedocs.io/en/develop/Users_Guide/tc-stat.html).

Hopefully that helps.

**Reply by YanJin-NOAA** [reply]
Hi John, 

Thanks very much for your detailed instruction. I will run the script by myself and learn to read the output file. 

Best
Yan

**Reply by YanJin-NOAA** [reply]
Hi John,

I am comparing the information in the performance diagram and the TC-Stat -rirw output file. 

The x-axis of the performance diagram is "Success Ratio", which can be calculated by ```1-FAR```, where FAR is False Alarm Rate, and it is included in the rirw output file. 

The y-axis of the performance diagram is "probability of detection, POD", which is defined by ```hit events / (hit events + miss events)```. In the rirw output file there are several columns containing "POD", such as "PODY", "PODY_NCL", "PODY_NCU","PODN", etc. I wonder which one should I use to plot the performance diagram? (I assume I should use "PODY" related column since PODN is almost 1) 

Thanks very much for your time and help.
Best,
Yan

**Reply by JohnHalleyGotway** [reply]
Yan, Appendix C of the MET User's Guide lists the equations for each of the statistics.

And I'm looking at this Performance Diagram reference:
https://www.cawcr.gov.au/projects/verification/Roebber/PerformanceDiagram.html

For the Y-axis, yes "PODY" is how MET writes the "Probability of Detection" (we put the Y at the end to differentiate between PODN).

The X-axis is a little tricky. There are 2 very similar stats...
- [False Alarm Rate](https://met.readthedocs.io/en/latest/Users_Guide/appendixC.html#probability-of-false-detection-pofd) (i.e. also called Probability of False Detection, or POFD)
- [False Alarm Ratio](https://met.readthedocs.io/en/latest/Users_Guide/appendixC.html#false-alarm-ratio-far) (called FAR in the MET output)

I see that the METviewer plotting script for [performance diagrams](https://github.com/dtcenter/METviewer/blob/main_v4.0/R_tmpl/performance.R_tmpl), plots **PODY versus 1-FAR** from MET. So that would be probability of detection versus success ratio (1 minus false alarm ratio (not rate)).

I find the whole "ratio" vs "rate" confusion very unfortunate!

**Reply by TaraJensen** [reply]
@JohnHalleyGotway @bikegeek @fisherhucar @TatianaBurek  - Is there anyway we can show Yan how to run 
METplotpy/metplotpy/plots/performance_diagram/performance_diagram.py on  ~/rirw_ctc_by_model.txt, even if it requires a few more lines to read the the output?  We already have the code to plot the Performance Diagram - it seems kinda crazy to reinvent the wheel.

**Reply by TaraJensen** [reply]
@YanJin-NOAA - If you can wait until next Tue/Wed, we will hopefully have a solution for you that doesn't require you to write your own Performance Diagram plotting script.

**Comment by JohnHalleyGotway** [comment]
@TaraJensen, as we discussed in the METplus meeting today, I wrote up this [development issue ](https://github.com/dtcenter/MET/issues/1960) to enhance TC-Stat to write full .stat output files that could be loaded into a database for use by METviewer. I tentatively assigned it to MET version 10.1.0 but did not assign it to a specific development cycle or particular developer or scientist. Please feel free to take a look and update/modify it however you see fit.

Note that the TC-Stat -job rirw and STAT-Analysis -job ramp both have this similar situation. They both write categorical output to custom, job-specific formats that cannot be easily loaded into a database for use by METviewer.

The development issue covers TC-Stat, not STAT-Analysis. But we could write up a similar one for STAT-Analysis at some point if needed. But the difficulty is in the details... deciding how to shoehorn job-specific info into the existing STAT header columns in an intuitive way.

**Comment by YanJin-NOAA** [comment]
Hi Tara,

I found a Python script to plot the performance diagram:
https://gist.github.com/djgagne/64516e3ea268ec31fb34

I plan to modify the code to 1) read the TC-Stat -job rirw output file, and
then 2) do the plotting. Currently the input data (model names, PODY, FAR)
is hard coded and saved to a dictionary. The next step is to read the file,
find the corresponding columns, and convert them to a dictionary or array.

The plot looks like this:

Thanks very much for your time and help. I feel it would be very helpful
for the verification team if we can visualize the result using MET/METplus.
Since TC-Stat has already generated the output file so we do need to do the
statistics by ourselves.

Best,
Yan

[image: image.png]

**Reply by YanJin-NOAA** [reply]
I just realized that the plot was not attached in the last post. (I replied the email and attached the plot in the email)
![image](https://user-images.githubusercontent.com/79869112/141662866-9500ace7-f0d2-4451-a7b1-3ff7ce6f00ed.png)
